union and intersect panic with slices of uncomparable values

I have this shortcode and I'm trying to get SLA % uptime seconds:

```
{{ $total_seconds := int (.Get "secs") }}
{{ $percents := slice 0.1 0.5 }}
{{ $percents := union (slice $percents) (slice (seq 1.0 1.0 99.0)) }}
{{ $percents := union (slice $percents) (slice 99.9 99.99 99.999 99.9999 99.99999) }}

Using {{ $total_seconds }} seconds as average.

<table>
{{ range $p := $percents }}

{{ $secs := sub 100.0 $p }}
{{ $secs := div $secs 100.0 }}
{{ $secs := mul $secs $total_seconds }}
<tr>
	<td>{{ $p }} %</td>
	<td>{{ $secs }}</td>
</tr>

{{ end }}
</table>
```

```
Started building sites ...
panic: runtime error: hash of unhashable type []interface {} [recovered]
        panic: runtime error: hash of unhashable type []interface {}

goroutine 111 [running]:
text/template.errRecover(0xc0427ff6c0)
        /usr/local/go/src/text/template/exec.go:140 +0x1f3
panic(0xb37ec0, 0xc04237cda0)
        /usr/local/go/src/runtime/panic.go:489 +0x2dd
github.com/gohugoio/hugo/tpl/collections.(*intersector).appendIfNotSeen(0xc0427feb78, 0xaf5ce0, 0xc0421df158, 0x97)
        /Users/bep/go/src/github.com/gohugoio/hugo/tpl/collections/collections.go:501 +0x1bb
github.com/gohugoio/hugo/tpl/collections.(*Namespace).Union(0xc04247a038, 0xaf5ce0, 0xc0421df248, 0xaf5ce0, 0xc0421df8d8, 0xc0421df8d8, 0x0, 0xc0427fec18, 0x410d5c)
        /Users/bep/go/src/github.com/gohugoio/hugo/tpl/collections/collections.go:565 +0x393
github.com/gohugoio/hugo/tpl/collections.(*Namespace).Union-fm(0xaf5ce0, 0xc0421df248, 0xaf5ce0, 0xc0421df8d8, 0x0, 0x0, 0x0, 0x0)
        /Users/bep/go/src/github.com/gohugoio/hugo/tpl/collections/init.go:122 +0x59
reflect.Value.call(0xb2df40, 0xc042040510, 0x13, 0xbed818, 0x4, 0xc0421df170, 0x2, 0x2, 0xf7da40, 0xb3dcc0, ...)
        /usr/local/go/src/reflect/value.go:434 +0x926
reflect.Value.Call(0xb2df40, 0xc042040510, 0x13, 0xc0421df170, 0x2, 0x2, 0xf784a0, 0xc0421f4b40, 0xaf5ce0)
        /usr/local/go/src/reflect/value.go:302 +0xab
text/template.(*state).evalCall(0xc0427ff640, 0xb665e0, 0xc0422b0300, 0x16, 0xb2df40, 0xc042040510, 0x13, 0xf781a0, 0xc0422f9ec0, 0xc04222429d, ...)
        /usr/local/go/src/text/template/exec.go:668 +0x526
text/template.(*state).evalFunction(0xc0427ff640, 0xb665e0, 0xc0422b0300, 0x16, 0xc0422f9ef0, 0xf781a0, 0xc0422f9ec0, 0xc04232fec0, 0x3, 0x4, ...)
        /usr/local/go/src/text/template/exec.go:536 +0x195
text/template.(*state).evalCommand(0xc0427ff640, 0xb665e0, 0xc0422b0300, 0x16, 0xc0422f9ec0, 0x0, 0x0, 0x0, 0xaf5ce0, 0xc0421df158, ...)
        /usr/local/go/src/text/template/exec.go:433 +0x6d4
text/template.(*state).evalPipeline(0xc0427ff640, 0xb665e0, 0xc0422b0300, 0x16, 0xc0422f0ff0, 0x0, 0x0, 0x97)
        /usr/local/go/src/text/template/exec.go:406 +0xf9
text/template.(*state).walk(0xc0427ff640, 0xb665e0, 0xc0422b0300, 0x16, 0xf78020, 0xc0422f6de0)
        /usr/local/go/src/text/template/exec.go:232 +0x561
text/template.(*state).walk(0xc0427ff640, 0xb665e0, 0xc0422b0300, 0x16, 0xf78380, 0xc0422f54a0)
        /usr/local/go/src/text/template/exec.go:240 +0x140
text/template.(*Template).execute(0xc0422eb6c0, 0xf6e000, 0xc04207a150, 0xb665e0, 0xc0422b0300, 0x0, 0x0)
        /usr/local/go/src/text/template/exec.go:195 +0x211
text/template.(*Template).Execute(0xc0422eb6c0, 0xf6e000, 0xc04207a150, 0xb665e0, 0xc0422b0300, 0xc0426426e0, 0x9)
        /usr/local/go/src/text/template/exec.go:178 +0x5a
html/template.(*Template).Execute(0xc0422f5440, 0xf6e000, 0xc04207a150, 0xb665e0, 0xc0422b0300, 0xc042527770, 0xc0421dd1c0)
        /usr/local/go/src/html/template/template.go:120 +0x93
github.com/gohugoio/hugo/tpl.(*TemplateAdapter).Execute(0xc04237c620, 0xf6e000, 0xc04207a150, 0xb665e0, 0xc0422b0300, 0xc0422b0300, 0x0)
        <autogenerated>:3 +0x7c
github.com/gohugoio/hugo/hugolib.renderShortcodeWithPage(0xf72340, 0xc04237c620, 0xc0422b0300, 0x0, 0x0)
        /Users/bep/go/src/github.com/gohugoio/hugo/hugolib/shortcode.go:721 +0xda
github.com/gohugoio/hugo/hugolib.renderShortcode(0xbeca4f, 0x2, 0xbed8c8, 0x4, 0xbeddf8, 0x4, 0xc042299ae0, 0x17, 0xc04253412d, 0x9, ...)
        /Users/bep/go/src/github.com/gohugoio/hugo/hugolib/shortcode.go:340 +0x63a
github.com/gohugoio/hugo/hugolib.prepareShortcodeForPage.func1(0xb307c0, 0xc04240c720, 0xc042527cc0, 0xfcd120)
        /Users/bep/go/src/github.com/gohugoio/hugo/hugolib/shortcode.go:261 +0xaf
github.com/gohugoio/hugo/hugolib.(*shortcodeHandler).executeShortcodesForDelta(0xc04232e9c0, 0xc042492000, 0x1b, 0xc042527de0)
        /Users/bep/go/src/github.com/gohugoio/hugo/hugolib/shortcode.go:406 +0xc5
github.com/gohugoio/hugo/hugolib.handleShortcodes(0xc042492000, 0xc042538000, 0x75, 0xd2, 0xc042286580, 0x72, 0xc042286080, 0x72, 0x80)
        /Users/bep/go/src/github.com/gohugoio/hugo/hugolib/hugo_sites.go:612 +0x1b1
github.com/gohugoio/hugo/hugolib.(*Site).preparePagesForRender.func1(0xc04207c780, 0xc042150530, 0xc04235f440, 0xc042402b20)
        /Users/bep/go/src/github.com/gohugoio/hugo/hugolib/hugo_sites.go:560 +0x24b
created by github.com/gohugoio/hugo/hugolib.(*Site).preparePagesForRender
        /Users/bep/go/src/github.com/gohugoio/hugo/hugolib/hugo_sites.go:591 +0xdd
```

Hugo Static Site Generator v0.26 windows/amd64 BuildDate: 2017-08-19T06:05:00+03:00

Example:

`uptime.md`:
```
---
title: "Uptime"
description: "Uptime SLA %"
#date: 2017-07-21T19:35:51+03:00
draft: false
categories: [projects]
tags: [projects]
---

# Month
{{< slauptime secs="2592000" >}} 

# Year
{{< slauptime secs="31536000" >}}

```

--- a/tpl/collections/collections.go
+++ b/tpl/collections/collections.go
@@ -298,8 +298,16 @@ func (ns *Namespace) Intersect(l1, l2 interface{}) (interface{}, error) {
 		case reflect.Array, reflect.Slice:
 			for i := 0; i < l1v.Len(); i++ {
 				l1vv := l1v.Index(i)
+				if !l1vv.Type().Comparable() {
+					return make([]interface{}, 0), errors.New("intersect does not support slices or arrays of uncomparable types")
+				}
+
 				for j := 0; j < l2v.Len(); j++ {
 					l2vv := l2v.Index(j)
+					if !l2vv.Type().Comparable() {
+						return make([]interface{}, 0), errors.New("intersect does not support slices or arrays of uncomparable types")
+					}
+
 					ins.handleValuePair(l1vv, l2vv)
 				}
 			}
@@ -609,6 +617,11 @@ func (ns *Namespace) Union(l1, l2 interface{}) (interface{}, error) {
 
 			for i := 0; i < l1v.Len(); i++ {
 				l1vv, isNil = indirectInterface(l1v.Index(i))
+
+				if !l1vv.Type().Comparable() {
+					return []interface{}{}, errors.New("union does not support slices or arrays of uncomparable types")
+				}
+
 				if !isNil {
 					ins.appendIfNotSeen(l1vv)
 				}

--- a/tpl/collections/collections_test.go
+++ b/tpl/collections/collections_test.go
@@ -360,10 +360,6 @@ func TestIntersect(t *testing.T) {
 		{[]int{1, 2, 4}, []int{3, 6}, []int{}},
 		{[]float64{2.2, 4.4}, []float64{1.1, 2.2, 4.4}, []float64{2.2, 4.4}},
 
-		// errors
-		{"not array or slice", []string{"a"}, false},
-		{[]string{"a"}, "not array or slice", false},
-
 		// []interface{} ∩ []interface{}
 		{[]interface{}{"a", "b", "c"}, []interface{}{"a", "b", "b"}, []interface{}{"a", "b"}},
 		{[]interface{}{1, 2, 3}, []interface{}{1, 2, 2}, []interface{}{1, 2}},
@@ -404,9 +400,18 @@ func TestIntersect(t *testing.T) {
 		{pagesVals{}, pagesVals{p1v, p3v, p3v}, pagesVals{}},
 		{[]interface{}{p1, p4, p2, p3}, []interface{}{}, []interface{}{}},
 		{[]interface{}{}, []interface{}{p1v, p3v, p3v}, []interface{}{}},
+
+		// errors
+		{"not array or slice", []string{"a"}, false},
+		{[]string{"a"}, "not array or slice", false},
+
+		// uncomparable types - #3820
+		{[]map[int]int{{1: 1}, {2: 2}}, []map[int]int{{2: 2}, {3: 3}}, false},
+		{[][]int{{1, 1}, {1, 2}}, [][]int{{1, 2}, {1, 2}, {1, 3}}, false},
+		{[]int{1, 1}, [][]int{{1, 2}, {1, 2}, {1, 3}}, false},
 	} {
 
-		errMsg := fmt.Sprintf("[%d]", test)
+		errMsg := fmt.Sprintf("[%d] %v", i, test)
 
 		result, err := ns.Intersect(test.l1, test.l2)
 
@@ -759,6 +764,10 @@ func TestUnion(t *testing.T) {
 		// errors
 		{"not array or slice", []string{"a"}, false, true},
 		{[]string{"a"}, "not array or slice", false, true},
+
+		// uncomparable types - #3820
+		{[]map[string]int{{"K1": 1}}, []map[string]int{{"K2": 2}, {"K2": 2}}, false, true},
+		{[][]int{{1, 1}, {1, 2}}, [][]int{{2, 1}, {2, 2}}, false, true},
 	} {
 
 		errMsg := fmt.Sprintf("[%d] %v", i, test)

